### ISSUE TITLE
Add getAll to MultiMap

### ISSUE THREAD

**Issue by nickersan** [hazelcast/hazelcast#3840]
Please add a `getAll` method to `MultiMap` to allow bulk retreval.  

The signature should be:

```
public MultiMap<K, V> getAll(Set<K> keys)
```

And should behave like `IMap.getAll(Set<K>)`

**Comment by inelpandzic** [comment]
Hello, I will be glad to take this issue.

**Comment by Holmistr** [comment]
@inelpandzic Fantastic! I'm assigning this issue to you. Shortly, I will get someone to give you some initial guidenance, tips etc. Looking forward to your contribution!

**Comment by tkountis** [comment]
Hi @inelpandzic thanks for looking into this.
I guess you could start by looking into how `IMap.getAll` does this and mimic this in Multimap.
This wont be the standard vanilla operation, it will rather rely on a small optimization that utilizes an OperationFactory which is send to the partition owners of the specified keys. Having said that, it should be pretty straight forward by following the IMap example. Feel free to poke me here or on gitter if you get stuck.

**Comment by inelpandzic** [comment]
Thanks a lot, @tkountis, that was exactly what I was going to do, see the `IMap.getAll` impl and follow along.

**Comment by inelpandzic** [comment]
Hi, @tkountis, for the return type of the `MultiMap.getAll` method  I'm thinking `Map<K, Collection<V>>`. What do you think?
So, the method would be:
```
public Map<K, Collection<V>> getAll(@Nonnull Set<K> keys)
```

**Comment by vbekiaris** [comment]
@inelpandzic method signature looks good to me.

**Comment by mmedenjak** [comment]
@vbekiaris @tkountis how about making it async as well?
`CompletionStage<Map<K, Collection<V>>> getAllAsync(@Nonnull Set<K> keys)`

**Comment by inelpandzic** [comment]
@mmedenjak if you agree, I would like to first finish the synchronous one, and after that, I could add the async one.

**Comment by vbekiaris** [comment]
There is no other async API on `MultiMap` yet. That said implementing the async variant should be fairly straightforward, return the internal `InvocationFuture` instead of blocking on `get()` (eg see https://github.com/hazelcast/hazelcast/blob/8d0effd4e72e273dcbca799a9cd46f41fc9c2e52/hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java#L261). Statistics handling also needs to be taken care of, but should also be straightforward.

**Comment by inelpandzic** [comment]
Ah yes, I can see, it is straightforward. I'm thinking then, after finishing this synchronous one, maybe to add async variants also for `get` and `put` and maybe some other method. Like there is on `IMap`.

**Comment by Holmistr** [comment]
I think what @mmedenjak is trying to say is, if we even want the sync version. Does it make much sense? We may ask "why not"? The answer would probably be that we want to keep the API as minimalistic as possible if there's no extreme demand for it.

Thoughts @vbekiaris @mmedenjak @jerrinot ?

**Comment by mmedenjak** [comment]
Yes, that's what I am asking. If we provide the asnyc version and someone wants the sync version, they can block on the returned value. If we provide the sync version and someone wants the async version, sorry. If we provide both versions, what is the added value of the sync method?

**Comment by mmedenjak** [comment]
On the other hand, if we had an API that only had async methods, it would be painful for people which want sync calls to use, they would need to constantly repeat the blocking pattern. So there are definite drawbacks. Personally I'd just add the `getAllAsync` method here. But I'm open to discussion here.

**Comment by inelpandzic** [comment]
Understand that, but would adding only `getAllAsync` somehow break the API consistency? Having only blocking `get` and only non-blocking `getAllAsync` feels a bit odd, what do you think?
For example, `IMap` has only `getAll`, and also has `get` and `getAsync`.

**Comment by mmedenjak** [comment]
It would. It's a result of a long history of adding methods without any agreement, and based on the fact that all `java.util.Map` methods don't need to return a future because everything happens locally. Personally, I prefer async methods and I would just add the async version. But let's first hear what others have to say.

**Comment by mmedenjak** [comment]
Maybe once we have project loom and once we switch to fibers, we simply won't care anymore about blocking and async :)

**Comment by Holmistr** [comment]
Hehe, we're getting of subject :D @mmedenjak Could we get a final and clear decision on which methods @inelpandzic should implement to keep him going?

**Comment by mmedenjak** [comment]
As I said, I'd prefer only an async method. I'll ask around to see if we can get more opinions.

**Comment by vbekiaris** [comment]
same here, async only seems fine to me.

**Comment by Holmistr** [comment]
@inelpandzic There you go, you have your answer :) `getAllAsync` only for the win!

**Comment by inelpandzic** [comment]
Well, until fibers, it's `getAllAsync` then :)

**Comment by Holmistr** [comment]
I hope you were not waiting for our confirmation @inelpandzic  :D Yes, we agree.

**Comment by Holmistr** [comment]
@c-attila In general, we never got it prioritized enough. MultiMap is not the most used datastructure (it was way behind Map, Queue, Topic, ...) and we just had bigger fish to fry. I would suggest you to reach through the Hazelcast Support and make a case for your usecase.

Disclaimer: I no longer work at Hazelcast.
